### PR TITLE
add `push_to_hub` to pipeline

### LINKED ISSUE #28857: add `push_to_hub( )` method when working with pipelines
### Feature request

it will be good to addition to the library to add a `push_to_hub( )` method when working with the pipeline method.

### Motivation

I have just finished up a blogpost about creating custom architectures, but i couldn't help but notice that when working with ` AutoConfig` or `AutoModelForxxx` or `PreTrainedModel`  have a `push_to_hub( )` methods alllowing them to easily be updated on the hub.
unfortunately, this isn't the case for the pipeline method, I found a possible work around by cloning the repo then using `save_pretrained( )` then pushing the changes, but this not optimised and takes lots of time especially when working with really large language models
check the documentation : 
https://huggingface.co/blog/not-lain/custom-architectures-with-huggingface#push-to-hub-%F0%9F%A4%97
<hr>

image
**VS**
image

### Your contribution

yes, I can help out with a PR

### PR DESCRIPTION
**Pull request by not-lain**
# What does this PR do?

this will add `push_to_hub` method to the pipelines allowing people to push their custom pipelines to the huggingface hub easily 

Fixes #28857 #28983

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@Narsil @Rocketknight1 

additional resources :
* https://huggingface.co/docs/transformers/custom_models (there is a `push_to_hub` method)
* https://huggingface.co/docs/transformers/en/add_new_pipeline (doesn't include a `push_to_hub` method)
* notebook: https://colab.research.google.com/drive/1yCET-FdkI4kHK1B-VUrrM7sKHMO3LkOg?usp=sharing

## TODO: 

- [x] fix documentation for `push_to_hub`
- [ ] ~~fix automap configuration for intial push (keeps adding a `user/repo--file.module`)~~ 
- [ ] update `save_pretrained` method (checkout [`PreTrainedModel`](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L2174) class for more info ) (enhancement)
- [x] add tests

### CODE REVIEW

**Comment by not-lain** [comment]
almost forgot #29004 will fix any problems with remote pipeline configuration for most of the cases (adds remote repo flags `user/repo--file.module` to the `custom-pipeline` field, leaving this as the final configuration inconsistency since this is related to the `auto_config` instead, adding extra [unnecessary remote flags](https://github.com/huggingface/transformers/pull/29172/commits/4be0c3510f834f696dc55ffd0ede1d340e7bf719#diff-441f558737166b045444da9c4be81f566b3d69054e8f20e288aed746a691fa61R938) , maybe we should add a if else there checking if the pipeline being pushed to the same original model repo.
a ruff estimation of the code in L938 of the same file could be like this : 
```python 
if self.model.config._name_or_path != repo_id :
  custom_object_save(self, save_directory)
```
let me know if you approve of this

**Comment by not-lain** [comment]
@ArthurZucker @Rocketknight1 after careful investigation it turns out that the extra flag is added due to the `AutoModelForxxx` and **NOT** the new `push_to_hub` method so i'm removing it from the todo list since it's irrelevant to this pull request.
reporoduction : 
https://colab.research.google.com/drive/1unFh3i5FyRRHcUO8Al7cLKkYXPhtr0lC?usp=sharing

**Comment by not-lain** [comment]
@Rocketknight1 I have updated the save_pretrained a little bit, the reason why i did this is that it's coupled with the push_to_hub method. I have done my part to cover as much ground as possible and this pull request is about the `push_to_hub` method so i will stop here.

since i don't know the repo that you want to test push to, i will that part to you to add them, this notebook will help you out when creating tests https://colab.research.google.com/drive/130IpVrScW8cNomEDY2Fa6-mA4_VrRmgT?usp=sharing

awaiting review

File: src/transformers/pipelines/base.py
Comment (Rocketknight1) [inline src/transformers/pipelines/base.py:948]: TODO not finished here!

File: src/transformers/pipelines/base.py
Comment (Rocketknight1) [inline src/transformers/pipelines/base.py:1262]: This bit confuses me - since you're inheriting `PushToHubMixin.push_to_hub`, `__doc__` should always be defined, right? I can see it's a copy of the same code for the other classes that inherit from `PushToHubMixin`, though, I'm just not sure why it's coded this way. Not a blocker, just a comment!

**Review by Rocketknight1 — APPROVED** [review]
Conditionally approving this PR too, except for one TODO, and the comment that I think we probably need to refactor / reconsider our model of custom pipelines and how they should be saved/loaded. See [this comment](https://github.com/huggingface/transformers/issues/28907#issuecomment-1959657870) for more.

File: src/transformers/pipelines/base.py
Comment (not-lain) [reply src/transformers/pipelines/base.py:1262]: since other people used this method to copy the docs i chose to use the same one as them, to stay in the same page as them, just to avoid straying too much from the norm

Comment (Rocketknight1) [reply src/transformers/pipelines/base.py:1262]: Yep, that's totally fine! I was just pointing out my own confusion, I guess

File: src/transformers/pipelines/base.py
Comment (not-lain) [reply src/transformers/pipelines/base.py:948]: since we are passing everything to the 
```diff
+ kwargs["safe_serialization"] = safe_serialization
+ self.model.save_pretrained(save_directory, **kwargs)
```
as kwargs i felt we should switch to a kwargs annotation
also yeh you are right, there is no need for deprecation or any changes, it already works perfectly as is, should i remove that comment ?

File: src/transformers/pipelines/base.py
Comment (not-lain) [reply src/transformers/pipelines/base.py:1262]: true that, i think the reason for them using this annotation is that they only need to change one method (the original one) to change the docs for all of the other classes using it.
meaning one changes all, which is a really nice approach

File: src/transformers/pipelines/base.py
Comment (Rocketknight1) [reply src/transformers/pipelines/base.py:948]: Yep - we don't want to leave unnecessary TODOs in the codebase!

Comment (not-lain) [reply src/transformers/pipelines/base.py]: removed the extra comment ✅

**Comment by not-lain** [comment]
cc @Rocketknight1 @ArthurZucker 
I added the new feature to the dynamic pipeline test thing
now it's finally ready for a review 🚀🤗

**Comment by not-lain** [comment]
cc @Rocketknight1 @ArthurZucker
any reviews on this one ?

**Comment by not-lain** [comment]
@Rocketknight1 friendly pinging you here.
Just wanted to say that the test that I added is working perfectly ✅

**Comment by Rocketknight1** [comment]
Sorry for the delay! I still feel like we might need to refactor our model for what custom pipelines actually do, but in the meantime this seems okay to add.

cc @amyeroberts for core maintainer review - this is basically a PR that adds `push_to_hub()` to custom pipelines. They already have a `save_pretrained()` method, so this just pushes the result of that.

We had some internal discussions about this, and at some point we might need to tackle the question of custom pipelines properly, including properly separating them from models (right now they're kind of attached at the hip to the model in their repo). Still, I think this fix is useful in the short-term!

**Comment by not-lain** [comment]
@Rocketknight1 
> including properly separating them from models

I do agree with you on this point and I do understand where you're coming from but I don't think that this is relevant much to this pr, even if we do seperate the model from the pipeline we still need the push_to_hub method.

imo we should create a separate issue for that

File: src/transformers/pipelines/base.py
Comment (amyeroberts) [inline src/transformers/pipelines/base.py:974]: We should have a `# Copied from` comment here as it's the same as the implementation in configuration_utils.py

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! 

Just one small comment

File: src/transformers/pipelines/base.py
Comment (not-lain) [reply src/transformers/pipelines/base.py:974]: well in `src\transformers\modeling_utils.py` in the `def save_pretrained` they didn't even import nor create a function to add the token to the kwargs.
IMO the `_set_token_in_kwargs` should be moved to the `PushToHubMixin` instead, this should help a lot to avoid changing every single class manually, for now I will settle in simply adding a comment since that enhancement is out of scope in this pr, do let me know if you approve of this, if so can you open another issue and tag me, I'll try to contribute to that

Comment (amyeroberts) [reply src/transformers/pipelines/base.py:975]: `_set_token_in_kwargs` is only defined in the config class. In fact, looking at it - we shouldn't need it here at all. This is a work around to account for the fact some models' config classes have their own `from_pretrained` method - but this isn't the case for pipelines

Comment (not-lain) [reply src/transformers/pipelines/base.py:975]: I see, thanks for the clarification

**Comment by not-lain** [comment]
Hi @amyeroberts 
Any reviews on this PR?

**Comment by amyeroberts** [comment]
@not-lain [Per this conversation](https://github.com/huggingface/transformers/pull/29172/files#r1551763830), the changes removing `_set_token_in_kwargs` should be done

File: src/transformers/pipelines/base.py
Comment (amyeroberts) [inline src/transformers/pipelines/base.py:917]: Having this isn't consistent with the logic in the method 

```suggestion
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating! 

Make sure that the input arguments are consistent with the logic and docstrings

File: src/transformers/pipelines/base.py
Comment (not-lain) [reply src/transformers/pipelines/base.py:917]: most `save_pretrained` methods pass the token in kwargs, even in here if I pass the token to the parameters I will need to set it up to kwargs again since the `push_to_hub` has the `token` as a parameter I think it will suffice to make it work.

fundamentally speaking it will be misleading to have the token in the `save_pretrained` method.

do let me know if you agree with this or if i should proceed with your changes

Comment (amyeroberts) [reply src/transformers/pipelines/base.py:917]: > do let me know if you agree with this or if i should proceed with your changes

I'm a bit confused. My proposed changes are to remove this as you say:

> fundamentally speaking it will be misleading to have the token in the save_pretrained method parameters.
> most save_pretrained methods pass the token in kwargs
> even in here if I pass the token to the parameters I will need to set it up to kwargs again

so we shouldn't have this as an argument in save_pretrained. What are the changes you're proposing?

Comment (not-lain) [reply src/transformers/pipelines/base.py:917]: oh, i have it as a parameter, i got confused a little, let me merge that

Comment (not-lain) [reply src/transformers/pipelines/base.py]: this is one of the very first times contributed to a project, so apologize if I got confused a lot, I merged your changes, also thanks for the feedback

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this feature!

**Pull request merged**
